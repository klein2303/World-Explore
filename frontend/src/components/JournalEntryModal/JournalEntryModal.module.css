.modal {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background-color: white;
    border-radius: 10px;
    box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.2);
    padding: 20px;
    width: 100%;
    max-width: 400px;
    z-index: 1000;
    overflow: hidden; /* Prevents content overflow */
}

.modalContent {
    position: relative;
    display: flex;
    flex-direction: column;
    gap: 15px;
    max-height: 90vh; /* Prevents modal from overflowing the screen */
    overflow-y: auto; /* Allows scrolling if the content is too tall */
}

.checkbox {
    display: flex;
    align-items: center;
}

.closeButton {
    position: absolute;
    top: 10px;  /* Adjust the distance from the top */
    right: 10px; /* Adjust the distance from the right */
    background: none;
    border: none;
    font-size: 1.5rem;
    cursor: pointer;
    padding: 0;
    margin: 0;
    z-index: 10; /* Ensure it stays above other elements */
}

.input {
    padding: 10px;
    font-size: 1rem;
    font-family: "Poppins", sans-serif;
    border: 1px solid #ccc;
    border-radius: 5px;
    background-color: #E4E1F7; /* Light purple background for input fields */
    color: #424242; /* Gray text */
}

.textArea {
    padding: 10px;
    font-size: 1rem;
    font-family: "Poppins", sans-serif;
    border: 1px solid #ccc;
    border-radius: 5px;
    height: 100px;
    background-color: #E4E1F7; /* Light purple background for textarea */
    color: #424242; /* Gray text */
    resize: none; /* Prevent user from resizing the textarea */
}

.rating {
    display: flex;
    align-items: center;
}

.ratingLabel {
    font-size: 1rem;
    color: #424242; /* Gray text */
    margin-right: 10px;
}

.star {
    font-size: 1.5rem;
    margin-right: 5px;
    cursor: pointer;
    color: transparent; /* No fill for inactive stars */
    stroke: #424242; /* Gray outline for stars */
    stroke-width: 15px; /* Thicker gray outline */
    transition: fill 0.2s ease, transform 0.2s ease; /* Smooth transition for fill */
}

.starActive {
    font-size: 1.5rem;
    margin-right: 5px;
    color: #E4E1F7; /* Purple fill for active stars */
    stroke: #424242; /* Keep gray outline for active stars */
    stroke-width: 15px; /* Thicker gray outline */
}

.star:hover,
.starActive:hover {
    stroke: #424242; /* Keep gray outline */
    fill: #E4E1F7; /* Stars turn purple on hover */
}

.submitButton {
    padding: 10px;
    font-size: 1rem;
    background-color: #E4E1F7; /* Purple background for the submit button */
    color: #424242; /* Gray text */
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s ease;
    font-weight: 500; /* Slightly bolder text */
}

.submitButton:hover {
    background-color: #d1cdf5; /* Slightly darker purple on hover */
}

label {
    font-size: 1rem;
    color: #424242; /* Gray text for public checkbox */
}

input[type="checkbox"] {
    margin-right: 10px;
    accent-color: #d1cdf5; /* Purple checkbox when checked */
}

.journalEntryTitle {
    color: #424242; /* Title color */
    text-align: center; /* Center the title */
}
