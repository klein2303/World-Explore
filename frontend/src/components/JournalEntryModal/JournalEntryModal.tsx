import { useState, useEffect, useRef } from "react";
import styles from "./JournalEntryModal.module.css";
import { FaStar } from "react-icons/fa";
import { removeQuotes } from "../../utils/utils";
import { gql, useMutation } from "@apollo/client";
import { useNavigate } from "react-router-dom";
import { JournalTypeWrite } from "../../types/JournalType";

interface JournalEntryModalProps {
    country: string;
    isOpen: boolean;
    onClose: () => void;
    onSubmit: (entry: JournalTypeWrite) => void;
}

const JournalEntryModal = ({ country, isOpen, onClose }: JournalEntryModalProps) => {
    const [title, setTitle] = useState<string>("");
    const [date, setDate] = useState<string>("");
    const [rating, setRating] = useState<number>(0);
    const [text, setText] = useState<string>("");
    const [isPublic, setIsPublic] = useState<boolean>(false);
    const [error, setError] = useState<string>("");

    const modalRef = useRef<HTMLDivElement>(null);
    const user = removeQuotes(sessionStorage.getItem("user")!);
    const navigate = useNavigate();

    const CREATE_REVIEW = gql`
        mutation addReview(
            $title: String!
            $date: String!
            $rating: Int!
            $text: String!
            $ispublic: Boolean!
            $profileid: String!
            $countryid: String!
        ) {
            addReview(
                title: $title
                date: $date
                rating: $rating
                text: $text
                ispublic: $ispublic
                profileid: $profileid
                countryid: $countryid
            ) {
                ispublic
            }
        }
    `;

    const [addReview] = useMutation(CREATE_REVIEW, {
        onCompleted: async () => {
            setError("");
            onClose();
            navigate(0);
        },
        onError: (error) => {
            setError(error.message);
        },
    });

    useEffect(() => {
        const focusableElements = modalRef.current?.querySelectorAll(
            'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])',
        );
        const firstElement = focusableElements?.[0] as HTMLElement;
        const lastElement = focusableElements?.[focusableElements.length - 1] as HTMLElement;

        const trapFocus = (e: KeyboardEvent) => {
            if (e.key === "Tab") {
                if (document.activeElement === firstElement && e.shiftKey) {
                    e.preventDefault();
                    lastElement?.focus();
                } else if (document.activeElement === lastElement && !e.shiftKey) {
                    e.preventDefault();
                    firstElement?.focus();
                }
            }
        };

        if (isOpen) {
            firstElement?.focus();
            document.addEventListener("keydown", trapFocus);
            document.body.classList.add("no-scroll"); // Add the class to disable scrolling
        } else {
            document.body.classList.remove("no-scroll"); // Remove the class to enable scrolling
        }

        return () => {
            document.removeEventListener("keydown", trapFocus);
            document.body.classList.remove("no-scroll"); // Clean up by removing the class
        };
    }, [isOpen]);

    const handleSubmit = () => {
        addReview({
            variables: {
                title: title,
                date: date,
                rating: rating,
                text: text,
                ispublic: isPublic,
                profileid: user,
                countryid: country,
            },
        });
    };

    return isOpen ? (
        <div
            className={styles.overlay}
            role="dialog"
            aria-modal="true"
            aria-labelledby="modal-title"
            aria-describedby="modal-description"
            ref={modalRef}>
            <aside className={styles.modal} role="dialog" aria-modal="true" aria-labelledby="journalModalTitle">
                <button onClick={onClose} className={styles.closeButton} aria-label="Close modal">
                    &times;
                </button>
                <div className={styles.modalContent} role="document">
                    <h2 className={styles.journalEntryTitle}>Write your journal entry for {country}</h2>
                    <input
                        type="text"
                        placeholder="Title"
                        value={title}
                        onChange={(e) => setTitle(e.target.value)}
                        className={styles.input}
                        aria-label="Title input"
                        maxLength={50}
                    />
                    <input
                        type="date"
                        value={date}
                        onChange={(e) => setDate(e.target.value)}
                        className={styles.input}
                        aria-label="Date input"
                    />
                    <textarea
                        placeholder="Relive the moment..."
                        value={text}
                        onChange={(e) => setText(e.target.value)}
                        className={styles.textArea}
                        aria-label="Journal entry text area"
                    />
                    {/* Star Rating System */}
                    <div className={styles.rating}>
                        <p className={styles.ratingLabel}>Rating:</p>
                        {[1, 2, 3, 4, 5].map((star) => (
                            <FaStar
                                key={star}
                                className={star <= rating ? styles.starActive : styles.star}
                                onClick={() => setRating(star)}
                                tabIndex={0} // Make the stars keyboard accessible
                                role="button"
                                aria-pressed={star <= rating} // Screen reader announcement for the current state
                                aria-label={`Rate ${star} star${star > 1 ? "s" : ""}`}
                                onKeyDown={(e) => {
                                    if (e.key === "Enter" || e.key === " ") setRating(star);
                                }}
                            />
                        ))}
                    </div>
                    <label className={styles.checkbox}>
                        <input
                            type="checkbox"
                            checked={isPublic}
                            onChange={() => setIsPublic(!isPublic)}
                            aria-label="Public checkbox"
                        />
                        Public
                    </label>
                    {error !== "" && <h5>{error}</h5>}
                    <button onClick={handleSubmit} className={styles.submitButton} aria-label="Submit journal entry">
                        Save your journal entry
                    </button>
                </div>
            </aside>
        </div>
    ) : null;
};

export default JournalEntryModal;
