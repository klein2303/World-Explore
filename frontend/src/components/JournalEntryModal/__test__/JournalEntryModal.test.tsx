import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { MockedProvider } from "@apollo/client/testing";
import gql from "graphql-tag"; // Import gql separately if not from Apollo Client
import JournalEntryModal from "../JournalEntryModal";
import userEvent from "@testing-library/user-event";

const mockClose = vi.fn();
const mockSubmit = vi.fn();
const mockNavigate = vi.fn();

vi.mock("react-router-dom", () => ({
    useNavigate: () => mockNavigate,
    useLocation: () => ({
        pathname: "/ExploreCountries",
    }),
}));

const mockSessionStorage = (() => {
    let store: Record<string, string> = {};
    return {
        getItem: (key: string) => store[key] || null,
        setItem: (key: string, value: string) => {
            store[key] = value;
        },
        clear: () => {
            store = {};
        },
    };
})();
Object.defineProperty(window, "sessionStorage", { value: mockSessionStorage });

const CREATE_REVIEW = gql`
    mutation addReview(
        $title: String!
        $date: String!
        $rating: Int!
        $text: String!
        $ispublic: Boolean!
        $profileid: String!
        $countryid: String!
    ) {
        addReview(
            title: $title
            date: $date
            rating: $rating
            text: $text
            ispublic: $ispublic
            profileid: $profileid
            countryid: $countryid
        ) {
            ispublic
        }
    }
`;

describe("JournalEntryModal", () => {
    beforeEach(() => {
        vi.clearAllMocks();
        sessionStorage.setItem("user", '"mockUserId"');
    });

    it("renders modal content correctly", () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <JournalEntryModal country="France" isOpen={true} onClose={mockClose} onSubmit={mockSubmit} />
            </MockedProvider>,
        );

        expect(screen.getByText("Write your journal entry for France")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Title")).toBeInTheDocument();
        expect(screen.getByLabelText("Date input")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Relive the moment...")).toBeInTheDocument();
        expect(screen.getByText(/Rating:/i)).toBeInTheDocument();
        expect(screen.getByLabelText("Public checkbox")).toBeInTheDocument();
        expect(screen.getByText(/Save your journal entry/i)).toBeInTheDocument();
    });

    it("traps focus within the modal when open", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <JournalEntryModal country="France" isOpen={true} onClose={mockClose} onSubmit={mockSubmit} />
            </MockedProvider>,
        );

        const closeButton = screen.getByRole("button", { name: /close modal/i });
        const titleInput = screen.getByPlaceholderText("Title");

        closeButton.focus();
        expect(document.activeElement).toBe(closeButton);

        await userEvent.tab();
        expect(document.activeElement).toBe(titleInput);

        await userEvent.tab();
        expect(document.activeElement).not.toBeNull(); // Ensure focus is cycling
    });

    it("handles star rating keyboard accessibility", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <JournalEntryModal country="France" isOpen={true} onClose={mockClose} onSubmit={mockSubmit} />
            </MockedProvider>,
        );

        const stars = screen.getAllByRole("button", { name: /Rate \d stars?/ });
        expect(stars.length).toBe(5);

        await userEvent.type(stars[2], "{Enter}");
        expect(stars[2]).toHaveAttribute("aria-pressed", "true");

        await userEvent.type(stars[3], "{Space}");
        expect(stars[3]).toHaveAttribute("aria-pressed", "true");
    });

    it("displays validation errors for missing fields", async () => {
        const mocks = [
            {
                request: {
                    query: CREATE_REVIEW,
                    variables: {
                        title: "",
                        date: "",
                        rating: 0,
                        text: "",
                        ispublic: false,
                        profileid: "mockUserId",
                        countryid: "France",
                    },
                },
                error: new Error("Validation failed"),
            },
        ];

        render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <JournalEntryModal country="France" isOpen={true} onClose={mockClose} onSubmit={mockSubmit} />
            </MockedProvider>,
        );

        fireEvent.click(screen.getByText(/save your journal entry/i));
        expect(await screen.findByText("Validation failed")).toBeInTheDocument();
    });

    it("clears error message after successful submission", async () => {
        const mocks = [
            {
                request: {
                    query: CREATE_REVIEW,
                    variables: {
                        title: "My Journey",
                        date: "2024-11-20",
                        rating: 5,
                        text: "A beautiful experience.",
                        ispublic: true,
                        profileid: "mockUserId",
                        countryid: "France",
                    },
                },
                result: {
                    data: {
                        addReview: {
                            ispublic: true,
                        },
                    },
                },
            },
        ];

        render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <JournalEntryModal country="France" isOpen={true} onClose={mockClose} onSubmit={mockSubmit} />
            </MockedProvider>,
        );

        // Fill the form
        await userEvent.type(screen.getByPlaceholderText("Title"), "My Journey");
        await userEvent.type(screen.getByLabelText("Date input"), "2024-11-20");
        await userEvent.type(screen.getByPlaceholderText("Relive the moment..."), "A beautiful experience.");
        fireEvent.click(screen.getAllByRole("button", { name: /Rate 5 stars/i })[0]);
        fireEvent.click(screen.getByLabelText("Public checkbox"));

        // Submit the form
        fireEvent.click(screen.getByText(/save your journal entry/i));

        // Wait for mutation to complete
        await screen.findByText("Write your journal entry for France");

        // Assert onClose is called
        expect(mockClose).toHaveBeenCalled();
    });
});
