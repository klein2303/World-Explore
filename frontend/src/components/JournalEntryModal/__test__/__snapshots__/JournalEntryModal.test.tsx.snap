// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JournalEntryModal > matches the snapshot when modal is open 1`] = `
<DocumentFragment>
  <div
    aria-describedby="modal-description"
    aria-labelledby="modal-title"
    aria-modal="true"
    class="_overlay_4dfb0c"
    role="dialog"
  >
    <aside
      aria-labelledby="journalModalTitle"
      aria-modal="true"
      class="_modal_4dfb0c"
      role="dialog"
    >
      <button
        aria-label="Close modal"
        class="_closeButton_4dfb0c"
      >
        Ã—
      </button>
      <div
        class="_modalContent_4dfb0c"
        role="document"
      >
        <h2
          class="_journalEntryTitle_4dfb0c"
        >
          Write your journal entry for Norway
        </h2>
        <input
          aria-label="Title input"
          class="_input_4dfb0c"
          maxlength="50"
          placeholder="Title"
          type="text"
          value=""
        />
        <input
          aria-label="Date input"
          class="_input_4dfb0c"
          type="date"
          value=""
        />
        <textarea
          aria-label="Journal entry text area"
          class="_textArea_4dfb0c"
          placeholder="Relive the moment..."
        />
        <div
          class="_rating_4dfb0c"
        >
          <p
            class="_ratingLabel_4dfb0c"
          >
            Rating:
          </p>
          <svg
            aria-label="Rate 1 star"
            aria-pressed="false"
            class="_star_4dfb0c"
            fill="currentColor"
            height="1em"
            role="button"
            stroke="currentColor"
            stroke-width="0"
            tabindex="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            />
          </svg>
          <svg
            aria-label="Rate 2 stars"
            aria-pressed="false"
            class="_star_4dfb0c"
            fill="currentColor"
            height="1em"
            role="button"
            stroke="currentColor"
            stroke-width="0"
            tabindex="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            />
          </svg>
          <svg
            aria-label="Rate 3 stars"
            aria-pressed="false"
            class="_star_4dfb0c"
            fill="currentColor"
            height="1em"
            role="button"
            stroke="currentColor"
            stroke-width="0"
            tabindex="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            />
          </svg>
          <svg
            aria-label="Rate 4 stars"
            aria-pressed="false"
            class="_star_4dfb0c"
            fill="currentColor"
            height="1em"
            role="button"
            stroke="currentColor"
            stroke-width="0"
            tabindex="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            />
          </svg>
          <svg
            aria-label="Rate 5 stars"
            aria-pressed="false"
            class="_star_4dfb0c"
            fill="currentColor"
            height="1em"
            role="button"
            stroke="currentColor"
            stroke-width="0"
            tabindex="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
            />
          </svg>
        </div>
        <label
          class="_checkbox_4dfb0c"
        >
          <input
            aria-label="Public checkbox"
            type="checkbox"
          />
          Public
        </label>
        <button
          aria-label="Submit journal entry"
          class="_submitButton_4dfb0c"
        >
          Save your journal entry
        </button>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;
